File: Leetcode/Python/1400. Construct K Palindrome Strings/solution.py
# https://leetcode.com/problems/construct-k-palindrome-strings/description

class Solution:
    def canConstruct(self, s: str, k: int) -> bool:
        if len(s) < k:
            return False
        if len(s) == k:
            return True

        letter_count = defaultdict(int)
        palindrome_count = 0

        for letter in s:
            letter_count[letter] += 1

        # A palindrome allows for one odd-count character, so the minimum number of characters required for a palindrome is an odd number
        for count in letter_count.values():
            if count % 2 != 0:
                palindrome_count += 1

        return palindrome_count <= k

File: Leetcode/Typescript/1400. Construct K Palindrome Strings/solution.ts
// https://leetcode.com/problems/construct-k-palindrome-strings/description

function canConstruct(s: string, k: number): boolean {
  if (s.length < k) {
    return false;
  }

  if (s.length === k) {
    return true;
  }

  const letterCounter = new Map<string, number>();
  let palindromeCount = 0;

  for (const letter of s) {
    letterCounter.set(letter, (letterCounter.get(letter) || 0) + 1);
  }

  // A palindrome allows for one odd-count character, so minimum number of palindrome is an odd-count number
  for (const count of letterCounter.values()) {
    if (count % 2 !== 0) {
      palindromeCount += 1;
    }
  }

  return palindromeCount <= k;
}
